src/java/test/org/apache/zookeeper/server/quorum/QuorumPeerMainTest.java

```diff
@@ -1861,6 +1862,76 @@ public void testFaultyMetricsProviderOnConfigure() throws Exception {
         Assert.assertTrue("complains about metrics provider MetricsProviderLifeCycleException", found);
     }
 
+    /**
+     * Test leader/leader compatibility with/without CloseSessionTxn, so that
+     * we can gradually rollout this code and rollback if there is problem.
+     */
+    @Test
+    public void testCloseSessionTxnCompatile() throws Exception {
```

Comment:
It's not a good practice to add parameter in the test class if only a single test is using it, have a test helper function like this seems better to me.

Reply:
Testing multiple things in a single unit test is not a good practice either. Also you could add more context to the method names like `testCloseSessionTxnCompatile_LeaderDisabled_FollowerEnabled()`.
Why not refactoring out to a separate test file then? This file is already quite overloaded: ~2000 lines of code.

From my experiences waiting for a refactoring ticket to fix these kind of issues doesn't lead to the resolution. Rectoring smaller chunks of code in PRs like this is essentially making progress. Following the existing bad practice is just increasing tech debt.

Reply:
When I was looking for places to put these kind of tests, this is the classes where I found most suitable, although I agree it's kind of overloaded.

This test is testing a single thing about compatibility, but that's debatable depends how much granularity you define a single purpose test. 

I'm not against of moving this out, create a separate test file, but considering we'll remove the closeSessionTxnEnabled option (we have started roll out this code to prod, and haven't seen problem, so it's likely I'll remove it here soon), that means this test will only include a single case leaderEnabled and followerEnabled, let me know if you think it's still worth to move out.

Reply:
Hm. Okay. I've created a pull request against your repo to visualize how I think these tests should be refactored: https://github.com/lvfangmin/zookeeper/pull/1

I understand this looks like overkill in terms of these tests being temporary, but I think it would be beneficial for the testing side on the long run.

I also would like to highlight that this is not the expected, final refactoring of `QuorumPeerMainTest` which will solve all of the related problems. This is just a small step that I was talking about which could be safely included into standards PRs like this and leads us to the ideal state of this big monster.

I'm not telling we should do it this way, just interested in your opinion.

Reply:
@eolivelli I think you're interested in this too.